### ISSUE TITLE
[feature]: Store the channel force close error message in the closed channels database

### ISSUE THREAD

**Issue by saubyk** [lightningnetwork/lnd#7696]
Currently the data stored in the closed channel database points to the type of channel closure e.g. `local force close` or `remote force close`, but not specifically the reason which may have triggered it.
Some examples can be:
- User initiated
- Low on reserve

Typically these reasons are printed on the console logs, which is hard to track for a user. If logged in the database, it can provide some additional insight to the user on the reason for force close, especially for the cases where the closure has been triggered automatically.

The `ClosedChannels` rpc should also be updated to include the reason in the `ChannelCloseSummary` response

**Comment by shaurya947** [comment]
Hey @saubyk this seems like a very useful addition! I recently had a channel force close by my own node while I was away and it was quite unclear as to why that happened.

I would like to give this a shot. I'll do some digging today to come up with a high-level plan of the required code changes, and share it when it's ready :slightly_smiling_face:

**Comment by shaurya947** [comment]
I spent a few hours looking through the code. There are many cases under which a channel may be marked as force closed in the DB:
- local force close
  - user initiated
  - link error
    - commitment point from remote doesn't match last unrevoked commit point
    - error when attempting to settle upstream HTLC
    - unable to accept new commitment (due to sig error)
- remote force close
- abandoning a channel
  - direct RPC call
  - as part of BatchOpenChannel
- funding errors
  - error before funding TX confirmed
  - timeout while waiting for funding TX conf

There may be more and/or I may have misinterpreted some cases. IMO the ones of interest here are:
- local force close
  - link error
    - commitment point from remote doesn't match last unrevoked commit point
    - error when attempting to settle upstream HTLC
    - unable to accept new commitment (due to sig error)
- remote force close

I would need some assistance from someone more experienced with the codebase regarding the following:

1. It's not obvious to me how to ascertain specific reasons for remote force close. I see that a lot of link failure messages are sent to the peer but I don't see where they may be processed and acted upon?
2. What would be the best way to pipe channel force close reason through the channel state machine? The force closing seems to happen inside `chain_arbitrator.go#ForceCloseContract()` where the force close request is queued. But the _actual_ marking of the channel as closed in the DB seems to happen within `channel_arbitrator.go` as various `ChainEvents` are observed. What would be a good design/architecture to bridge the two instances?

**Comment by saubyk** [comment]
cc: @yyforyongyu

**Comment by yyforyongyu** [comment]
I'd recommend take a look at this [transition flow](https://github.com/lightningnetwork/lnd/blob/master/contractcourt/briefcase.go#L133) to understand it more. Notice that we've already have `CloseType` and `ChanStatusFlags` to give us some insights about the FC. What's nice to have is to somehow record the state steps that have happened from calling [this method](https://github.com/lightningnetwork/lnd/blob/master/contractcourt/channel_arbitrator.go#L806).

**Comment by shaurya947** [comment]
Thanks for the links @yyforyongyu ! I'll try to absorb all of this over the next few days and come back with a plan and/or more questions :slightly_smiling_face:

**Comment by shaurya947** [comment]
Hey @yyforyongyu I looked through the state transitions. I've got a few more questions for you. Sorry that this is a lengthy comment. I've underlined the 4 exact spots that are the actual questions:

<ins>Could you please confirm that there is no feasible way to extract a specific reason for a remote force closure?</ins> I say this because the path for registering a remote force closure is as follows:
  1. https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/chain_watcher.go#L623
  2. https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/chain_watcher.go#L758
  3. https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/chain_watcher.go#L1127
  4. https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/channel_arbitrator.go#L566
  5. https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/channel_arbitrator.go#L2684
  6. https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/channel_arbitrator.go#L2742
  7. https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/channel_arbitrator.go#L1403
  8. https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/channel_arbitrator.go#L892
  9. https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/channel_arbitrator.go#L893

So basically, our observer that's watching for on-chain events recognizes a broadcast of the commitment TX by the remote party, and in response to that notification, our state machine transitions to `StateContractClosed` and we don't have any actionable info as to why the remote party chose to broadcast their TX. Is that accurate? If so, I can stop worrying about the remote force closure and focus my attention only on local force closure.

---

As for local force closures, there seem to be 2 paths invoking the `userTrigger`:
- `CloseChannel` RPC request:
  1. https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/rpcserver.go#L2438
  2. https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/rpcserver.go#L2529
  3. https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/chain_arbitrator.go#L1067
  4. https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/chain_arbitrator.go#L1091
  5. https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/channel_arbitrator.go#L2838
  6. https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/channel_arbitrator.go#L2853
- Link failure in `brontide.go`:
  1. https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/peer/brontide.go#L3080
  2. https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/peer/brontide.go#L3103
  3. ...Rest of steps same as above

<ins>Does it seem correct that a link-failure force close uses the exact same trigger as a user-requested force close? Even if we were to record the states within the arbitrator's log, there doesn't seem to be a way to distinguish between these two paths unless we use different triggers/states.. Is that correct?</ins>

---

There also seem to be 2 paths invoking `chainTrigger`:
- new block notification being received
  - there are no on-chain actions required -> state remains as `StateDefault` so we can ignore this case https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/channel_arbitrator.go#L849
  - there are on-chain actions required (HTLC timeout or settle) -> state changes to `StateBroadcastCommit` so we need to consider this case https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/channel_arbitrator.go#L872
- resolution signal being received https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/channel_arbitrator.go#L2815
<ins>As far as I can tell a resolution signal can only be received AFTER `StateContractClosed` which means we can ignore this case because our channel has already been marked closed. Is that accurate?</ins>

---

All that being said, my plan is as follows. <ins>As a final question, I would love your feedback on this plan</ins>:
- Decide on one of the following two options (so we have knowledge of the triggers in addition to knowledge of the states):
  - divide the `StateBroadcastCommit` state into further substates as necessary such as `StateBroadcastCommitUserTrigger`, `StateBroadcastCommitLinkFailureTrigger` (if we wish to distinguish between link failure vs user initiated), and `StateBroadcastCommitChainTrigger` (could be broken down further depending on exact reason such as HTLC timeout etc.)
  - OR when storing the states within the Arbitrator, also store the trigger responsible for the state change (could be a struct or a 2-byte flag with first byte being state and second byte being trigger)
- Extend the `ArbitratorLog` interface to add a `States` function that returns a slice of all the states up to `CurrentState` https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/briefcase.go#L60
- When committing a new state using the `CommitState` function, instead of overriding a single state, we instead append to a slice of states and store that new slice. We do the append iff the new state is not the same as the last state so that the slice doesn't grow unbounded with a number of redundant states. https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/briefcase.go#L543
- We also update the `CurrentState` function accordingly to return the last element of the slice https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/briefcase.go#L502
- When marking a channel as closed with the initiator being local, we examine the log slice here https://github.com/lightningnetwork/lnd/blob/ac3e7be9a063bdb559855f3b5af15e0323878450/contractcourt/channel_arbitrator.go#L2661 and based on the triggers within the slice, we can mark the channel as one of following:
  - user initiated
  - link error
  - HTLC timeout or on-chain settlement required

**Comment by shaurya947** [comment]
Hey @yyforyongyu just wanted to make sure this didn't disappear from your radar :slightly_smiling_face:

**Comment by ziggie1984** [comment]
Solid analysis @shaurya947 👍 

## General Thoughts

Before jumping into your concept I think it's worth discussing the information gain we would like to have here. Meaning that if we implement this kind of force-close reason I think it should be as specific as possible. That being said I think the most interesting reasons we would link to track here is the LinkFailure and especially the detailed reason as outlined in the `linkFailureReport`. Would be cool to funnel this report into the db somehow. Just differentiating between Link-Failure and Onchain-Action required is not really worth the hustle (not counting for the user-triggerd because IMO thats not quite relevant, you probably know why this was needed).

To give an example in case of a Force-Close because of a Link-Failure:

Let's say we Force-Close because of a triggered `ErrCommitSyncRemoteDataLoss` is way different compared to `ErrBelowChanReserve` which would both be labeled as `StateBroadcastCommitLinkFailureTrigger ` in your example which in my view brings not a lot of benefit tbh.

## Now coming to your questions and proposal:

> Could you please confirm that there is no feasible way to extract a specific reason for a remote force closure? 

This is more fine granular as it seems at first glance. Meaning that you are right that we cannot really be sure why our Peer broadcasted his Commitment Transaction, but there are cases where we can say this with much more certainty than expected. Let's say we suffer a DataLoss which would trigger the Peer to trigger a Force-Close, now if we receive the RemoteCommitment we can be very sure that we anticipated this Remote FC.
Same think might also be said in cases where we mark a Channel as inactive ans send an Error to the Peer. 
https://github.com/lightningnetwork/lnd/blob/6833df52368aca81109614ccd38cfc66913ea781/peer/brontide.go#L3120
I think it might be a good idea to mark the channel here with the reason of the failure, so maybe not borked, but at least the last error why we had to wipe the channellink : https://github.com/lightningnetwork/lnd/blob/6833df52368aca81109614ccd38cfc66913ea781/peer/brontide.go#L3095
Especially these Failures we send to our Peer do not lead to Force-Closes in most of the cases (at least LND does not Force Close on most of the errors for now, but other implementation might do so) but essentially marking the channel as unusable and finally needing userintervention to Force-Close this channel. By logging this error in the DB, we might even get more insights why a user had to Force-Close the channel manually (because it was unusable anyways)

So I think we can do better in the remote closure case as well.

> Does it seem correct that a link-failure force close uses the exact same trigger as a user-requested force close?

Yes I think you analysis is right, LinkFailures and User-Requested FCs use the same trigger. This kind of touches my comment about logging the real error of the link failure which could gain us some insights in both cases because LinkFailures could lead to User-Requested FCs eventually. Maybe instead of looking into detailing the trigger would it maybe be worthwhile to enhance the `ChannelStatus` enum with more detailed cases, because it's a bitarray it can be combined so it seems handy ?

> As far as I can tell a resolution signal can only be received AFTER StateContractClosed which means we can ignore this case because our channel has already been marked closed. Is that accurate?

Yes I agree your analysis is right here.

> All that being said, my plan is as follows. As a final question, I would love your feedback on this plan:

So not sure about splitting up the trigger/state logic could definitely be a way to increase information gain but would still leave a lot of cases undifferentiated?

Not sure whats the real benefit to map all the states when the `advanceState` function is triggered, the potential states a channel will run through seems pretty obvious to me and I am not sure how they would enhance our knowledge about the Force-Close reason ?


I think we should definitly have a differentiation between the following errors and they should be logged and visual represented in the `ClosedChannels GRPC`.

1. Why did the User Trigger the Force-Close => was the Channel Link failed anyways and we had no other choicee (DataLoss by the peer etc...)
2. When failing because of a LinkFailure => what was the main Failure all about
3. When Force-Closing the Channel because of the `chaintrigger` what was the exact reason: HTLC-Timeout (on local) HTLC-Timeout (on remote), Incoming HTLC running into the Broadcastdelta, I guess these were all?


To sum it up very good analysis so far, consider my comments with care because I am not that familiar with the codebase and these are just ideas I mentioned above so feel free to discard.

**Comment by shaurya947** [comment]
Hey @ziggie1984 thanks so much for taking the time to read and respond :slightly_smiling_face: 

> the most interesting reasons we would like to track here is the LinkFailure and especially the detailed reason as outlined in the linkFailureReport

^ I fully agree with this.

> I think we can do better in the remote closure case as well.

^ Your example of tracing cause (local data loss) and effect (remote force close) over time and linking them together makes sense. However, to keep scope manageable, I will likely focus on the more straightforward cases first, and leave remote force closures for follow-up tasks.

So I think for the first PR, we could target the following outcomes:
- user initiated (not differentiating between user willfully initiating vs link failure forcing user-FC eventually yet, could be done in follow-ups)
- link error (below three are the only specific errors that cause force close as far as I can tell)
  - ErrSyncError
  - ErrInvalidUpdate
  - ErrInvalidCommitment 
- chain trigger
  - outgoing HTLC timeout
  - incoming HTLC timeout
  - ... any other types identifiable via the `ChainActionMap` for local/remote

Do you think that would be a good starting point? And then we can build on top of that iteratively?

> Not sure whats the real benefit to map all the states when the advanceState function is triggered, the potential states a channel will run through seems pretty obvious to me and I am not sure how they would enhance our knowledge about the Force-Close reason ?

I was trying to figure out a way to pass data like trigger details etc. to the `ChannelArbitrator` so that we can reference those details when we actually mark the channel as closed here:
https://github.com/lightningnetwork/lnd/blob/0005adaac8031625bde0e181dfa24d13acceb081/contractcourt/channel_arbitrator.go#L2612

My initial idea was to create more substates depending on the trigger, and then to log all the states. But now that we are being more granular, maybe we need another way to make that data available. You have a lot more experience with this code than me, do you have any ideas? :smile: Should I just create new fields within the `ChannelArbitrator` struct to store these new data?

**Comment by shaurya947** [comment]
Hey @ziggie1984 thanks for the suggestions.

I tried first to add new fields to the `OpenChannel` struct but I ran into two issues with that:
1. I couldn't make the update in memory from `brontide.go` (link failure) or `rpcserver.go` (user requested FC) because the `arbChannel` type (that wraps the in-memory `OpenChannel` instance used by the `ChannelArbitrator`) is not exported https://github.com/lightningnetwork/lnd/blob/753af11ed636c57b88689bb9a8d5efaae59e4847/contractcourt/chain_arbitrator.go#L256
2. Because I couldn't make the update in memory, I would've had to make the update in the DB (in the `openChannelBucket`), and then later read from it. This felt a bit wrong because this info didn't seem to belong in this bucket.

After a lot of deliberation, I decided to make use of the `ChannelArbitrator`'s `ArbitratorLog`. In addition to the `ArbitratorState`, there are other things logged here as well like contract resolutions, commit sets, etc. which made it seem like the right place for our new data re: local force close info. https://github.com/lightningnetwork/lnd/blob/753af11ed636c57b88689bb9a8d5efaae59e4847/contractcourt/briefcase.go#L96

I made a proof of concept commit here on my fork: https://github.com/shaurya947/lnd/commit/c3df58b53ff8808af9d2b21deef06e359b900ca8

Let me know what you think :slightly_smiling_face:

**Comment by shaurya947** [comment]
Added two more commits here https://github.com/shaurya947/lnd/commits/force-close-reason

Now we have a working itest to verify the user-initiated case :tada: 

I will clean this up next week and make a proper PR.

**Comment by ziggie1984** [comment]
Looking forward for the real PR ⚡️, left some comments in your concept, the concept looks promising :)

**Comment by shaurya947** [comment]
Looks like the comments are still present on the dangling commit so feel free to respond to my queries there or on https://github.com/lightningnetwork/lnd/pull/7787 thank you :slightly_smiling_face:

**Comment by Chinwendu20** [comment]
Hello @saubyk, I would like to continue with the good work done by @shaurya947 as regards this issue.

**Comment by saubyk** [comment]
> Hello @saubyk, I would like to continue with the good work done by @shaurya947 as regards this issue.

Assigned to you. Thanks for taking the initiative. 🙏

**Comment by saubyk** [comment]
This issue is deprioritized from 0.18. Will pick it up in the next

**Comment by litbot-9000** [comment]
Closing as duplicate of #10317 which consolidates requests for better force close information. Unique details from this issue have been summarized there.
